rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,452 @@
+[2]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix
+[3]: https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix
+
+In general, the format for writing documentation strings is **not formally specified**.
+
+Among the formats encountered in the wild, the one used in `nixpkgs/lib` is the only one intended to be rendered as part of an API documentation, via the nixdoc third-party tool, whose syntax has not been standardized.
+
+### Impossible to differentiate from internal comments
+
+The lack of a formal definition of a doc-comment also means there is no reliable way to distinguish them from internal comments, which makes it impossible to access from doc-comments from tooling. Futhermore it makes it very hard to generate accurate and complete reference documentation.
+
+### References to the problems above
+
+This curated link collection highlights the Nix ecosystem's inconsistencies, a primary focus of this RFC.
+
+#### nixpkgs - comment examples
+
+- [lib/attrsets](https://github.com/NixOS/nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/lib/attrsets.nix)
+- [trivial-builders](https://github.com/NixOS/nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/pkgs/build-support/trivial-builders/default.nix)
+- [stdenv/mkDerivation](https://github.com/NixOS/nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/pkgs/stdenv/generic/make-derivation.nix)
+- [nixos/lib/make-disk-image](https://github.com/NixOS/nixpkgs/blob/5323fbf70331f8a7c47f1b4f49841cf74507f77f/nixos/lib/make-disk-image.nix)
+
+# Design
+[design]: #detailed-design
+
+In the following we give a comprehensive overview to our decision that we've made. Detailed arguments for and against every decision can be found in the [Decisions](#Decisions) section
+
+## CommonMark
+
+**CommonMark according to [RFC 72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md) is the content of all doc-comments.**
+
+Adopting CommonMark as the content for all doc-comments brings the benefits of widely accepted and understood documentation format in tech projects, while maintaining profitability and consistency within the Nix ecosystem by aligning with existing [NixOS/RFC-72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md).
+
+## `/** */` is the doc-comment format
+
+The decision to use /** to start a doc-comment ensures a unique distinction from regular comments while still allowing seamless writing without IDE or editor support. A single choice provides the best developer experience and simplifies tooling.
+
+## Placement
+
+**The placement describes the relationship between doc-comments and the expression they are documenting.**
+
+The following rules apply in descending order of precedence:
+
+- Doc-comments are placed before the documentable node. Only whitespace is allowed in between. ([Examples](#basic-examples)) 
```

Comment:
I think it is somehow unclear that what is *documentable*. Examples here shows that they are "expressions", which can be evaluated and has a associated value. But can we document some language constructs that is not an expression? For example, lambda formal?

```nix
{
/** how to document `a`? */
a
} : 1
```

We actually do have this kind of documentation in nixpkgs, e.g. https://github.com/NixOS/nixpkgs/blob/fe1e9d524b7ac9fac5f8be3ea2a7d177a5a4b4dc/lib/options.nix#L68

Reply:
With our POC retrieving this kind of documentation is already possible. The challenge is finding the right formulation. We can conclude that it is not strictly expressions that are documentable.

I think i'll just clarify what `documentable nodes` are (by the state of this RFC).
We can extend the list in the future if we find specifications are missing.

- expressions
- lambda formals
- bindings

This should cover about 99% of all documentation use cases and (as far as i can tell) sufficient for nixpkgs.